### ISSUE TITLE
Drastically reduce cardano-node memory usage by setting RTS params

### ISSUE THREAD

**Issue by weebl2000** [input-output-hk/daedalus#2734]
Since most heap in use by cardano-node is long-lived stuff, passing the `+RTS -c -RTS` should result in roughly half the memory usage that cardano-node is currently using.

It might be possible to reduce memory usage further by setting `-F1.5`, but I'm not sure about that one.

Also: CPU usage is usually high when running Daedalus. I think it would be wise to set `-I0` (no idle GCs) or setting `-Iw600`, idle GC only every 10 minutes, for the cardano-node process.

For reference:

https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/runtime_control.html#rts-options-to-control-the-garbage-collector

![image](https://user-images.githubusercontent.com/2385844/140534716-fa71b4f9-c79f-4a73-9d81-1b34f3352251.png)

Personally I'd maybe also increase the generations.

But long story short: I think cardano-node memory usage can be drastically reduced by simply setting some RTS params. Currently I do not know how to pass RTS parameters to cardano-node from Daedalus.

**Comment by weebl2000** [comment]
I've also been testing using "-xn" (non-copying collector), but at least for GHC 8.10.4 it does not seem 100% stable and can cause crashes. Though following GHC release 8.10.6 there have been improvements made to the non-copying collector.

https://downloads.haskell.org/ghc/8.10.7/docs/html/users_guide/8.10.6-notes.html

Fix a crash on the non-moving GC due to incorrect logic for calculating the header size for array write barries (#19715)

**Comment by dmitrii-gaico** [comment]
Hello! Thank you for the report and suggestions!
Currently we are preparing a release which includes [Improvements](https://github.com/input-output-hk/daedalus/pull/2735) to reduce RAM usage. Would be glad to hear your feedback after the official builds are out. 
Thank you!

**Comment by CherryDT** [comment]
I think the question of how to control these arguments is valid though, even with these improvements. After I got the DbLocked issue fixed and it was all working well for a while, I now see it suddenly crashing at random intervals and in the log it shows `heap exhausted`. I fixed it now by changing the launcher code to set `-M10000M` instead of `-M6553M`, but it would be great if this could be controlled from the outside somehow.

With the `-M10000M` it seems to be stable now, and it uses ~7 GB RAM, although I do wonder why I need >6.5G of RAM for the node now...? It always used a lot of RAM but it didn't crash with heap exhaustion before. I have 32GB of RAM so it's OK for me to allocate 10GB, but what would happen if someone had only 8GB RAM?

(I read that those settings somehow also scale based on processor cores... For the record, I have an AMD Ryzen 7 PRO 4750U with 8 cores and 2 threads, i.e. 16 vitual cores.)

**Comment by weebl2000** [comment]
> I think the question of how to control these arguments is valid though, even with these improvements. After I got the DbLocked issue fixed and it was all working well for a while, I now see it suddenly crashing at random intervals and in the log it shows `heap exhausted`. I fixed it now by changing the launcher code to set `-M10000M` instead of `-M6553M`, but it would be great if this could be controlled from the outside somehow.
> 
> With the `-M10000M` it seems to be stable now, and it uses ~8 GB RAM, although I do wonder why I need >6.5G of RAM for the node now...? It always used a lot of RAM but it didn't crash with heap exhaustion before. I have 32GB of RAM so it's OK for me to allocate 10GB, but what would happen if someone had only 8GB RAM?
> 
> (I read that those settings somehow also scale based on processor cores... For the record, I have an AMD Ryzen 7 PRO 4750U with 8 cores and 2 threads, i.e. 16 virtual cores. It's a ThinkPad T14s 1st Gen laptop. Is it somehow possible that the new settings in 4.5.0 improve it for many users but for my particular configuration they make it _worse_...?)
> 
> ![image](https://user-images.githubusercontent.com/6010463/142624548-cedc10a4-f293-4a0a-9f7c-896f6e84152b.png)
> 
> You can see it here starting out with 4 GB, then 4.7 GB soon thereafter, then later jumping to 7 GB, and then finally to 8 GB. Before the change, it would crash when it did that jump from 4.7 GB to 7 GB.

Hi CherryDT. This is unicorn advice, but can you do the same and try just these RTS params: "-I0 -w -xn"

I think this will DRASTICALLY reduce memory usage.

**Comment by CherryDT** [comment]
@weebl2000 Thank you! I tried your advice and changed it to `nodeConfig.rtsOpts = ['-I0', '-w'];` in `resources\app\dist\main\index.js` and the behavior is definitely different now, more of a continuous ramp than those fixed jumps.

The final result seems definitely _better_ although the improvement was not as drastic as I hoped. After the same 20 minutes that the last screenshot showed, it was at 5.3 GB with this configuration.

It quickly went to 3.6 GB initially, then while the output was stuck on `Opened lgr db` it creeped up to 4.3 GB, and then once it started responding (and Daedalus went to displaying my wallet info) it went up to 5.3 GB quickly and stayed there for a while (node was already synced). After a few minutes, I had 4x the error `[CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.93 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")` in the log, and it went up by 1 GB to 6.3 GB. Note that in the previous test, I had the exact same thing happening and it went from 7 GB to 8 GB then. **I think this is a memory leak!** In fact, while I was writing this, the same thing happened a second time, and it want up by another 1 GB from 6.3 GB to 7.3 GB!

![image](https://user-images.githubusercontent.com/6010463/142626993-27c49d61-ea58-4a5c-a39b-a121631fce01.png)

**Comment by weebl2000** [comment]
It's not really a memory leak. The problem is in GHC MEGABLOCK (1MB) are never returned to OS. The "-w" is a simple single threade mark/sweep GC but I think it is perfect for cardano-node!

**Comment by weebl2000** [comment]
> @weebl2000 Thank you! I tried your advice and changed it to `nodeConfig.rtsOpts = ['-I0', '-w'];` in `resources\app\dist\main\index.js` and the behavior is definitely different now, more of a continuous ramp than those fixed jumps.
> 
> The final result seems definitely _better_ although the improvement was not as drastic as I hoped. After the same 20 minutes that the last screenshot showed, it was at 5.3 GB with this configuration.
> 
> It quickly went to 3.6 GB initially, then while the output was stuck on `Opened lgr db` it creeped up to 4.3 GB, and then once it started responding (and Daedalus went to displaying my wallet info) it went up to 5.3 GB quickly and stayed there for a while (node was already synced). After a few minutes, I had 4x the error `[CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.93 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")` in the log, and it went up by 1 GB to 6.3 GB. Note that in the previous test, I had the exact same thing happening and it went from 7 GB to 8 GB then. **I think this is a memory leak!** In fact, while I was writing this, the same thing happened a second time, and it want up by another 1 GB from 6.3 GB to 7.3 GB!
> 
> ![image](https://user-images.githubusercontent.com/6010463/142626993-27c49d61-ea58-4a5c-a39b-a121631fce01.png)

You can try one more thing, might not make things better but:

+RTS -w -G2 -A128m -n1m -C0 -I0.3 -Iw3600 -O4000m -RTS

**Comment by weebl2000** [comment]
> Hm, can you help me understand, how is this not a memory leak if it will consume more and more memory every time this pipe error occurs, and then eventually crash because it reaches some limit?

Yeah, it's not really a cardano-node problem. It's a GHC problem. Megablocks in GHC are never returnedf to the OS in 8.10.7. In the newest GHC 9.2.1 they have made efforts to return memory to the OS.

![image](https://user-images.githubusercontent.com/2385844/142628315-d164566d-a790-425b-883b-b8d425941893.png)

If you want to chat more about this, not sure if you have a matrix account, but you can reach me on @weebl:matrix.weebl.me 

I tried updating cardano-node to 9.2.1 but it was dependency hell of stuff that still relies on GHCS base 4.15 or lower.

Let me know what `+RTS -N2 -w -A128m -n1m -I0.3 -Iw600 -O3500m -I0 -RTS` does for your memory usage. Make sure to wait at least 10-15 minutes before checking!

**Comment by weebl2000** [comment]
> Hm, can you help me understand, how is this not a memory leak if it will consume more and more memory every time this pipe error occurs, and then eventually crash because it reaches some limit?

To add to this, if you create 1000 objects that are each in their own megablock but actually very small. It will never return 1000 MB to the OS.

**Comment by CherryDT** [comment]
Hm whatever the root cause may be, it's strange that I get those pipe errors in the log at random intervals and then the memory usage jumps up. I left it running now, and with your `-I0 -w` setting there is no limit at which it would crash, so now the memory reached 15 GB... I'm assuming it will continue to grow and eat everything my computer has. :|

I guess my question is how come that these pipe errors appear and not only that but also cause this memory thing every time? Since a pipe error doesn't seem related to a memory-management dependency, I still get the feeling there is something cardano-node should do differently here.

I'll try your new configuration now.

**Comment by weebl2000** [comment]
> Hm whatever the root cause may be, it's strange that I get those pipe errors in the log at random intervals and then the memory usage jumps up. I left it running now, and with your `-I0 -w` setting there is no limit at which it would crash, so now the memory reached 15 GB... I'm assuming it will continue to grow and eat everything my computer has. :|
> 
> I guess my question is how come that these pipe errors appear and not only that but also cause this memory thing every time? Since a pipe error doesn't seem related to a memory-management dependency, I still get the feeling there is something cardano-node should do differently here.
> 
> I'll try your new configuration now.

Yeah -I0 is evil for running cardano-node in Daedalus. You can reduce -Iw600 to Maybe -Iw300 (every 5 minutes idle GC instead of every 10 minutes).

**Comment by weebl2000** [comment]
> Hm whatever the root cause may be, it's strange that I get those pipe errors in the log at random intervals and then the memory usage jumps up. I left it running now, and with your `-I0 -w` setting there is no limit at which it would crash, so now the memory reached 15 GB... I'm assuming it will continue to grow and eat everything my computer has. :|
> 
> I guess my question is how come that these pipe errors appear and not only that but also cause this memory thing every time? Since a pipe error doesn't seem related to a memory-management dependency, I still get the feeling there is something cardano-node should do differently here.
> 
> I'll try your new configuration now.

Oh btw. Is your cardano node fully synced? Or is it still busy syncing?

**Comment by CherryDT** [comment]
It's fully synced. Just adding new blocks.

I was about to open this issue with `cardano-node` before you mentioned it's not supposed to be their issue - this shows you more detail:

> **Internal/External**
> *External*   otherwise.
> 
> **Area**
> *Other*   Any other topic (Delegation, Ranking, ...).
> 
> **Summary**
> See https://github.com/input-output-hk/daedalus/issues/2734#issuecomment-974035789 and https://github.com/input-output-hk/daedalus/issues/2734#issuecomment-974051523 for how I discovered this.
> 
> I have a ThinkPad T14s 1st Gen laptop with 32 GB RAM and Windows 10. I'm experiencing this problem when using Daedalus, but it seems it's an issue in the node itself, that's why I'm opening this issue here.
> 
> I discovered two issues with the release of Daedalus 4.5.0:
> 
> * After a while the node would crash with `heap exhausted`. After patching `-M6553M` to `-M10000M` that seemed solved at first.
> * At random intervals, I see 4 `MuxIOException`s in the log, and at the same time RAM usage jumps up by 1 GB. (This happens also if I use `-I0 -w` instead of `-H4G -M10000M -c70` as suggested by @weebl2000.)
> 
> **Steps to reproduce**
> Steps to reproduce the behavior:
> 1. Have a similar system configuration to mine, I guess
> 2. Start Daedalus
> 3. Monitor cardano-node.exe's RAM usage and the node.log file
> 
> **Expected behavior**
> No ever-increasing RAM usage
> 
> **System info (please complete the following information):**
> - OS Name: Windows
> - OS Version 10.19043.1288
> - Node version cardano-node 1.30.1 - mingw32-x86_64 - ghc-8.10
> - CLI version  cardano-cli 1.30.1 - mingw32-x86_64 - ghc-8.10
> 
> **Screenshots and attachments**
> ```
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:57:06.03 UTC] Chain extended, new tip: 59a4c4365c5c24716a50f3b6d0749ef2af18ad9466e99bf1c23020f7c82c6e77 at slot 45760333
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:57:18.13 UTC] Chain extended, new tip: ae4b7175d108460b98682afc26eeed0f6625a6019b80cb4962eb4d451fdd85dd at slot 45760345
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 12:57:21.67 UTC] Block fits onto some fork: 27789ae08cf2bcebab42856e33a3e2f16c2e6c3f3ee91c5ff7bd887764f96467 at slot 45760347
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 12:57:21.79 UTC] Valid candidate 27789ae08cf2bcebab42856e33a3e2f16c2e6c3f3ee91c5ff7bd887764f96467 at slot 45760347
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:57:21.79 UTC] Switched to a fork, new tip: 27789ae08cf2bcebab42856e33a3e2f16c2e6c3f3ee91c5ff7bd887764f96467 at slot 45760347
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.93 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.93 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.94 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:57:21.94 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:57:36.81 UTC] Chain extended, new tip: 4a901b4648e5369fe50a1d0665196ba418e3b9e316925c532c8170f6d11fcb87 at slot 45760364
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:57:51.50 UTC] Chain extended, new tip: e30607b454ffb11a2637ea2c22cce483640cdc417cff374155fe068bc47db583 at slot 45760376
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:59:42.53 UTC] Chain extended, new tip: 5c4c6d3e002eedb63d1ce640be04e5fb05f108b8d389ad625a61fe762bd30919 at slot 45760489
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:59:45.66 UTC] Chain extended, new tip: df2b13e1452277378104d9cbefddbe4dd29d3a1508fa487d423a8fc2ca785aac at slot 45760493
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 12:59:47.17 UTC] Block fits onto some fork: 6da702753dd3c675be4def81aa9180bd0842ad6f496cd96c83f2680baad52cd2 at slot 45760494
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 12:59:47.27 UTC] Valid candidate 6da702753dd3c675be4def81aa9180bd0842ad6f496cd96c83f2680baad52cd2 at slot 45760494
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:59:47.27 UTC] Switched to a fork, new tip: 6da702753dd3c675be4def81aa9180bd0842ad6f496cd96c83f2680baad52cd2 at slot 45760494
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:59:47.34 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:59:47.37 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:59:47.39 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 12:59:47.39 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 12:59:51.71 UTC] Chain extended, new tip: 4856144fbe42e8550c657323fc83b14607b9d2ec3bf5c557fd107a46e09e9c46 at slot 45760499
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:00:49.52 UTC] Chain extended, new tip: d610411706dee449a685a7797f6dca1a3c75f18764843817a1bb99e9ad4193bb at slot 45760557
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:01:01.95 UTC] Chain extended, new tip: 51a9886108484d5462db7f472846f6e96cd10b20eeb497c2ebf48363551a3ee1 at slot 45760569
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:01:04.77 UTC] Chain extended, new tip: 0f639fcdda23c4278f8c388de71a5e91ba96efde626fdfaf01b973a94a438515 at slot 45760572
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:01:54.90 UTC] Chain extended, new tip: b4fa78c8bca03d66b2b9707d690e5ccb9023a874383a552ca45639743755a8a3 at slot 45760622
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:02:41.57 UTC] Chain extended, new tip: b1b8aa5beae3749fab815a9b40ace84570bb0d066795d8b2482f88b073045d72 at slot 45760668
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:03:06.40 UTC] Chain extended, new tip: 4ed3a8669a292b240c4ed41855af9ed0e431e8639b49d1e15c74c73c5a33e934 at slot 45760693
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:03:32.96 UTC] Chain extended, new tip: d60f33f67d8107f7624cb9a6cf291450d1deb7a0abd41e46d7d74d5ac07aaf52 at slot 45760720
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:04:24.75 UTC] Chain extended, new tip: 6cf0d9077a1be7968ec18ca5720086e75aee81906a4367ff6bc86c9b344a5f28 at slot 45760772
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 13:04:25.13 UTC] Block fits onto some fork: d7b29b971c49f74631c144e074d549dad0f2dced1d835c92debbdcb5383d52c8 at slot 45760772
> [CHE-T14S:cardano.node.ChainDB:Info:25] [2021-11-19 13:04:25.24 UTC] Valid candidate d7b29b971c49f74631c144e074d549dad0f2dced1d835c92debbdcb5383d52c8 at slot 45760772
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:04:25.24 UTC] Switched to a fork, new tip: d7b29b971c49f74631c144e074d549dad0f2dced1d835c92debbdcb5383d52c8 at slot 45760772
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 13:04:25.35 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 13:04:25.36 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 13:04:25.36 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.LocalErrorPolicy:Error:63] [2021-11-19 13:04:25.36 UTC] IP LocalAddress "\\\\.\\pipe\\cardano-node-mainnet.24496.0" ErrorPolicyUnhandledApplicationException (MuxError (MuxIOException withIODataPtr (readHandle): resource vanished (Die Pipe wurde beendet.)) "(readHandle errored)")
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:04:45.84 UTC] Chain extended, new tip: d5dbfdd91f8398f0fed50b2405112df2c5573aef9beb7d22983871512dbfb847 at slot 45760792
> [CHE-T14S:cardano.node.ChainDB:Notice:25] [2021-11-19 13:04:49.56 UTC] Chain extended, new tip: 21c2fe21daba7a892c62493eb9391d17eac2652a22287c586d6d2bbc2a5e063f at slot 45760797
> ```
> 
> ![image](https://user-images.githubusercontent.com/6010463/142627599-8a715f1b-c080-413d-8968-dfbd72194e18.png)
> 
> Point 1 was when the node was done initializing the database and became responsive to TCP connections.
> 
> **But points 2, 3, 4 and 5 were four instances of this pipe error!**
> 
> This was with `-I0 -w`, but as mentioned, the same behavior occurs with `-H4G -M10000M -c70`, just with faster "jumps" instead of those "ramps".
> 
>

**Comment by CherryDT** [comment]
OK the result of using `+RTS -N2 -w -A128m -n1m -I0.3 -Iw300 -O3500m -I0 -RTS` is that the initial memory usage after DB loading is 4.98 GB instead of 5.25 GB as it was before, and now the pipe errors don't seem to be connected to the RAM jumps anymore... The RAM still increases at irregular intervals but this time by 0.5 GB and not 1 GB, and I cannot see a correlation to the pipe errors anymore...

Overall I can say the climb rate is lower now though.

Here is the performance after 15 minutes:

![image](https://user-images.githubusercontent.com/6010463/142631572-400b85ee-93a7-4328-94d7-37ffb78cb9da.png)

Actually, looking at the log again, I get the feeling that the pipe errors are - while more prominent - just a consequence of the _previous_ log entry, which is always `Switched to a fork, new tip`... so maybe _the fork thing_ prompts memory operations that triggered those jumps before...

Still, this can't be right. Doesn't matter whether it's a dependency or not that causes it - it didn't happen in 4.4.1, it happens in 4.5.0, and it eats memory (even leads to regular crashes unless you mess with the code, which isn't something a regular user would do), so I call it a memory leak introduced in this version. If it walks like a duck, squaks like a duck... you know.

**Comment by weebl2000** [comment]
You're welcome. I'm not enitrely sure if this is actually memory in use or just "occupied" by cardano-node but that is easily swapped out. I don't think it matters much, but you might try one more extra param:

`+RTS -N2 -w -A64m -n1m -I0.3 -Iw600 --disable-delayed-os-memory-return -RTS`

**Comment by CherryDT** [comment]
I don't know if it may be easily swapped out or not, but at least my OS didn't yet swap it out, because I had physical memory usage of 24 GB and when I closed the wallet again it went down to 14 GB now, so...

I'll try your new suggestion.

**Comment by weebl2000** [comment]
> OK the result of using `+RTS -N2 -w -A128m -n1m -I0.3 -Iw300 -O3500m -I0 -RTS` is that the initial memory usage after DB loading is 4.98 GB instead of 5.25 GB as it was before, RAM usage still increases at irregular intervals but this time by 0.5 GB and not 1 GB, and I cannot see a correlation to the pipe errors anymore...
> 
> Overall I can say the climb rate is lower now though.
> 
> Here is the performance after 15 minutes:
> 
> ![image](https://user-images.githubusercontent.com/6010463/142631572-400b85ee-93a7-4328-94d7-37ffb78cb9da.png)
> 
> Actually, looking at the log again, I get the feeling that the pipe errors are - while more prominent - just a consequence of the _previous_ log entry, which is always `Switched to a fork, new tip`... so maybe _the fork thing_ prompts memory operations that triggered those jumps before...
> 
> Still, this can't be right. Doesn't matter whether it's a dependency or not that causes it - it didn't happen in 4.4.1, it happens in 4.5.0, and it eats memory (even leads to regular crashes unless you mess with the code, which isn't something a regular user would do), so I call it a memory leak introduced in this version. If it walks like a duck, quacks like a duck... you know.
> 
> (But still thanks for your help, this means that now I can at least use the wallet for half an hour at a time without it crashing.)

I think I might have passed the wrong params before, try my latest suggestion.

**Comment by CherryDT** [comment]
Still happens, but a bit slower.

![image](https://user-images.githubusercontent.com/6010463/142637683-65f706f6-dc15-44b5-b21e-b701e0ffa1b2.png)

**Comment by weebl2000** [comment]
> Still happens, but a bit slower.
> 
> ![image](https://user-images.githubusercontent.com/6010463/142637683-65f706f6-dc15-44b5-b21e-b701e0ffa1b2.png)

Oh wow, that's actually very high usage. Hmm. I've been running a cardano-node on a 4 GB machine without any swapping. Not sure if it's just trying to use all the memory there is...

**Comment by weebl2000** [comment]
> Still happens, but a bit slower.
> 
> ![image](https://user-images.githubusercontent.com/6010463/142637683-65f706f6-dc15-44b5-b21e-b701e0ffa1b2.png)

How sure are you that params are passed to the cardano-node?

**Comment by CherryDT** [comment]
Hmm, I looked at the command line of the process now, and there was `+RTS` and `-RTS` **twice** each time. I did replace the arguments of `-H4G -M6553M -c70` with what you sent, but since Daedalus already passed them between `+RTS <...> -RTS`, it means now with the new arguments that you wrote after `-I0 -w`, which also had `+RTS <...> -RTS`, that it was now `+RTS +RTS <...> -RTS -RTS`.

I fixed that now and tried again, so now it looks like this:

![image](https://user-images.githubusercontent.com/6010463/142640888-f39aa045-f790-4611-a059-089aa51b2624.png)

...but the result is still the same:

![image](https://user-images.githubusercontent.com/6010463/142643090-ac6e57d0-940d-4f90-ae29-92a0022da231.png)

**Comment by weebl2000** [comment]
> Hmm, I looked at the command line of the process now, and there was `+RTS` and `-RTS` **twice** each time. I did replace the arguments of `-H4G -M6553M -c70` with what you sent, but since Daedalus already passed them between `+RTS <...> -RTS`, it means now with the new arguments that you wrote after `-I0 -w`, which also had `+RTS <...> -RTS`, that it was now `+RTS +RTS <...> -RTS -RTS`.
> 
> I fixed that now and tried again, so now it looks like this:
> 
> ![image](https://user-images.githubusercontent.com/6010463/142640888-f39aa045-f790-4611-a059-089aa51b2624.png)
> 
> ...but the result is still the same:
> 
> ![image](https://user-images.githubusercontent.com/6010463/142643090-ac6e57d0-940d-4f90-ae29-92a0022da231.png)

I'd be willing to try this next week on Linux. I actually have a the Daedalus source code checked out and have ran yarn nix:mainnet etc. Can you make a branc with your changes so I can pull it and try the same thing on LInux?

**Comment by CherryDT** [comment]
I patched it in the already-transpiled JS code, but it corresponds to https://github.com/input-output-hk/daedalus/blob/e9623ecc8919891c7e68b618ce675d178dd433d1/source/main/cardano/CardanoWalletLauncher.js#L145

**Comment by weebl2000** [comment]
> I patched it in the already-transpiled JS code, but it corresponds to this code location:
> 
> https://github.com/input-output-hk/daedalus/blob/e9623ecc8919891c7e68b618ce675d178dd433d1/source/main/cardano/CardanoWalletLauncher.js#L145

If I forget to check this on Linux, can you remind me next week? I've been playing around with GHC RTS params before because I think the memory+CPU usage of cardano-node when running in Daedalus is completely stupid. I have one PR there which is about to be merged which should improve CPU usage when using cardano-node in a *nix build

https://github.com/input-output-hk/cardano-node/pull/3349

**Comment by CherryDT** [comment]
Okay! Thank you.

**Comment by CherryDT** [comment]
By the way, you asked if it is actually "used" or can be easily paged out.

It seems the ever-growing part can be easily paged out, but it's still a problem because my virtual memory can't have infinite size either.

I forced it to page out all of its memory and then watched it page back in what's needed, and this way I was able to get the working set down from 16 GB to 3 GB. But those other 13 GB still count towards the total commit charge on my system and would cause out-of-memory errors eventually if they'd keep growing.

![image](https://user-images.githubusercontent.com/6010463/142647684-361ab29c-aa71-4bd6-96be-9775d3198447.png)

**Comment by weebl2000** [comment]
I will definitely try next week on Linux. I am able to run cardano-node (plain) without about 3-4GB of heap usage.

**Comment by weebl2000** [comment]
```
[2021-11-22T23:01:17.279Z] [electron-connect] [server] started electron process: 940684
[2021-11-22T23:01:17.280Z] [electron-connect] [server] server created and listening on 30080
App threw an error during load
Error: /nix/store/vbr3p4y2azrgwq2s7h2hbip8x06w50z5-glibc-2.32-24/lib/libc.so.6: version `GLIBC_2.33' not found (required by /home/weebl/git/daedalus/node_modules/usb/build/Debug/usb_bindings.node)
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1846)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1138:18)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5:1846)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12684)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at bindings (/home/weebl/git/daedalus/dist/main/index.js:28782:48)
    at Object../node_modules/usb/usb.js (/home/weebl/git/daedalus/dist/main/index.js:211076:112)
(node:940684) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `electron --trace-deprecation ...` to show where the warning was created)
A JavaScript error occurred in the main process
Uncaught Exception:
Error: /nix/store/vbr3p4y2azrgwq2s7h2hbip8x06w50z5-glibc-2.32-24/lib/libc.so.6: version `GLIBC_2.33' not found (required by /home/weebl/git/daedalus/node_modules/usb/build/Debug/usb_bindings.node)
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1846)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1138:18)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5:1846)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12684)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at bindings (/home/weebl/git/daedalus/dist/main/index.js:28782:48)
    at Object../node_modules/usb/usb.js (/home/weebl/git/daedalus/dist/main/index.js:211076:112)

```

I'm getting errors running the nix build from latest master. I'm guessing they need to update nix packages first to support GLIBC 2.33

**Comment by weebl2000** [comment]
@dmitrii-gaico I think this was implemented in Daedalus 4.8 - do we want to close this or keep it open for looking into further RTS param tuning? I'd be open to help out with that.

**Comment by dmitrii-gaico** [comment]
Thank you all for the participation and feedback. Daedalus 4.8.0 had been released and includes the RTS Mode feature. For now we decided to stick with "-c" option only even it is showing not the best ram usage reduction. However based on our testing results "-c" is the safest option for the Node stability. Other flag combinations showed better ram usage improvements but unfortunately the Node started to work unstable. 
If you are able to find a combination which is showing better ram usage and Node stability is not affected please feel free to open a new card and share your findings. It would be much appreciated.
For now I'm closing the current ticket.
